core-bundle/src/Migration/CommandCompiler.php

```diff
@@ -13,193 +13,54 @@
-                case str_starts_with($sql, 'CREATE FULLTEXT INDEX '):
-                    $return['ALTER_ADD'][md5($sql)] = $sql;
-                    $order[] = md5($sql);
-                    break;
-
-                case str_starts_with($sql, 'DROP INDEX'):
-                    $return['ALTER_CHANGE'][md5($sql)] = $sql;
-                    $order[] = md5($sql);
-                    break;
-
-                case preg_match('/^(ALTER TABLE [^ ]+) /', $sql, $matches):
-                    $prefix = $matches[1];
-                    $sql = substr($sql, \strlen((string) $prefix));
-                    $parts = array_reverse(array_map('trim', explode(',', $sql)));
-
-                    for ($i = 0, $count = \count($parts); $i < $count; ++$i) {
-                        $part = $parts[$i];
-                        $command = $prefix.' '.$part;
-
-                        switch (true) {
-                            case str_starts_with($part, 'DROP '):
-                                $return['ALTER_DROP'][md5($command)] = $command;
-                                $order[] = md5($command);
-                                break;
-
-                            case str_starts_with($part, 'ADD '):
-                                $return['ALTER_ADD'][md5($command)] = $command;
-                                $order[] = md5($command);
-                                break;
-
-                            case str_starts_with($part, 'CHANGE '):
-                            case str_starts_with($part, 'RENAME '):
-                                $return['ALTER_CHANGE'][md5($command)] = $command;
-                                $order[] = md5($command);
-                                break;
-
-                            default:
-                                $parts[$i + 1] .= ','.$part;
-                                break;
-                        }
-                    }
-                    break;
-
-                default:
-                    throw new \RuntimeException(sprintf('Unsupported SQL schema diff: %s', $sql));
-            }
-        }
-
-        $this->checkEngineAndCollation($return, $order, $fromSchema, $toSchema);
-
-        $return = array_filter($return);
-
-        // HOOK: allow third-party developers to modify the array (see #3281)
-        if (isset($GLOBALS['TL_HOOKS']['sqlCompileCommands']) && \is_array($GLOBALS['TL_HOOKS']['sqlCompileCommands'])) {
-            foreach ($GLOBALS['TL_HOOKS']['sqlCompileCommands'] as $callback) {
-                $return = System::importStatic($callback[0])->{$callback[1]}($return);
-            }
-        }
+        // Get a list of SQL commands that adjust the engine and collation options
+        $engineAndCollationCommands = $this->compileEngineAndCollationCommands($fromSchema, $toSchema);
 
-        $this->commands = $return;
-        $this->commandOrder = $order;
+        return array_unique([...$diffCommands, ...$engineAndCollationCommands]);
```

Comment:
Do we need this unique check?
Could there be (in the future) the same query twice like a scenario like `do a, do b, revert a, do c, do a`?